service/storage_service.cc

```diff
@@ -1019,6 +1031,43 @@ future<> storage_service::raft_bootstrap(raft::server& raft_server) {
     }
 }
 
+future<> storage_service::raft_update_supported_features(raft::server& raft_server) {
+    // TODO: Can we avoid a read barrier? I don't think we can just check
+    // what's persisted on disk and then skip if it matches our feature set
+    // because something different might be already committed (e.g. we committed
+    // a larger feature set, did not apply it locally but later decided to
+    // roll back). Perhaps we could persist a local flag, outside raft,
+    // which indicates whether the committed state corresponds to our local,
+    // applied state. Each node is responsible for its own features in raft
+    // so this optimization might work.
+
+    co_await raft_server.read_barrier(&_abort_source);
+
+    const auto& state_machine_features = _topology_state_machine._topology.supported_features;
+    const auto features_set_v = _feature_service.supported_feature_set();
+    const std::set<sstring> features_set{features_set_v.begin(), features_set_v.end()};
+
+    while (true) {
+        const auto features_it = state_machine_features.find(raft_server.id());
+        if (features_it != state_machine_features.end() && features_it->second == features_set) {
+            break;
+        }
+
+        const sstring features_string = gms::feature_service::from_feature_set(features_set);
+
+        auto guard = co_await _group0->client().start_operation(&_abort_source);
+        topology_mutation_builder builder(guard.write_timestamp(), raft_server.id());
```

Comment:
I wonder if this is safe, since the topology coordinator might also be updating this entry concurrently (?) Consider the situation: the topology coordinator has decided to remove this node from the cluster. It removes the `topology` row for this server. However concurrently we're re-adding the row by the command we send here.

(this can only happen if we're still a member of group 0 - perhaps the coordinator, when removing a node, should start by removing it from group 0, and only then removing the `topology` entry for this node)

Perhaps after taking the group 0 guard, we should check if we're still a member of the cluster (by looking if we have an entry in the state machine). This should be done while holding the guard for atomicity.